aws-transfer-server/src/test/java/software/amazon/transfer/server/UpdateHandlerTest.java

```diff
@@ -205,6 +205,13 @@ public void stabilizationFailures() {
 
     private ProgressEvent<ResourceModel, CallbackContext> updateServerAndAssertStatus(
             ResourceHandlerRequest<ResourceModel> request, String postUpdateState, OperationStatus operationStatus) {
+        // Create a copy to avoid the update handler mutating the original.
+        request = request.toBuilder()
+                .previousResourceState(
+                        request.getPreviousResourceState().toBuilder().build())
+                .desiredResourceState(
+                        request.getDesiredResourceState().toBuilder().build())
+                .build();
```

Comment:
So, this is actually simpler as this `request.toBuilder().build();` because toBuilder() makes a copy to start with. So, if all you do is copy the previous values it is not doing anything useful.

Reply:
Does `toBuilder` make a deep copy?

We specifically need to avoid mutation of the `ResourceModel` (the previous and desired states). I was originally just creating copies of those two states and setting the request to use these copies.

However, some tests do not directly create the `ResourceModel`, and only get the model via the request object that is returned by a helper function. So these tests would end up with the copy of the `ResourceModel` that gets mutated.

Reply:
Hmm, how does that work? The SDK v2 code is all immutable classes. How do they get mutated?

Reply:
It has been a while since I worked this code, perhaps it does do this as I remember it is basically mocking the server responses so the exact object references are passed around.

Reply:
The request object is immutable, but the `ResourceModel` is mutable. I think the update handler mutates it when it merges tags.

This is when I was experimenting with making it return null instead of an empty list for tags. What I observed is that `ResourceModel` would have `null` for tags when it was created by the test, but after the update handler was invoked, the tags on the `ResourceModel` would instead be an empty list.

Reply:
Well, then ignore me. You are correct and it needs copying.